pkg/controller/elasticsearch/remotecluster/remoteca/controller.go

```diff
@@ -0,0 +1,289 @@
+	// that may have been existed in the past but should not exist anymore.
+	actualRemoteCertificateAuthorities, err := getActualRemoteCertificateAuthorities(ctx, r.Client, localClusterKey)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	results := &reconciler.Results{}
+	// Create or update expected remote CA
+	for remoteEsKey := range expectedRemoteCertificateAuthorities {
+		// Get the remote Elasticsearch cluster associated with this remote CA
+		remoteEs := &esv1.Elasticsearch{}
+		if err := r.Client.Get(remoteEsKey, remoteEs); err != nil {
+			if errors.IsNotFound(err) {
+				// Remote cluster does not exist, skip it
+				continue
+			}
+			return reconcile.Result{}, err
+		}
+		accessAllowed, err := isRemoteClusterAssociationAllowed(r.accessReviewer, localEs, remoteEs, r.recorder)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// if the remote CA exists but isn't allowed anymore, it will be deleted next
+		if !accessAllowed {
+			continue
+		}
+		delete(actualRemoteCertificateAuthorities, remoteEsKey)
+		results.WithResults(createOrUpdateCertificateAuthorities(ctx, r, localEs, remoteEs))
+		if results.HasError() {
+			return results.Aggregate()
+		}
+	}
+
+	// Delete existing but not expected remote CA
+	for toDelete := range actualRemoteCertificateAuthorities {
+		log.V(1).Info("Deleting remote CA",
+			"local_namespace", localEs.Namespace,
+			"local_name", localEs.Name,
+			"remote_namespace", toDelete.Namespace,
+			"remote_name", toDelete.Name,
+		)
+		results.WithError(deleteCertificateAuthorities(ctx, r, localClusterKey, toDelete))
+	}
+	return results.WithResult(association.RequeueRbacCheck(r.accessReviewer)).Aggregate()
+}
+
+func caCertMissingError(cluster types.NamespacedName) string {
+	return fmt.Sprintf("Cannot find CA certificate cluster %s/%s", cluster.Namespace, cluster.Name)
+}
+
+// getExpectedRemoteCertificateAuthorities returns all the remote cluster keys for which a remote ca should created
+// The CA certificates must be copied from the remote cluster to the local one and vice versa
+func getExpectedRemoteCertificateAuthorities(
+	ctx context.Context,
+	c k8s.Client,
+	associatedEs *esv1.Elasticsearch,
+) (map[types.NamespacedName]struct{}, error) {
+	span, _ := apm.StartSpan(ctx, "get_expected_remote_ca", tracing.SpanTypeApp)
+	defer span.End()
+	expectedRemoteClusters := make(map[types.NamespacedName]struct{})
+
+	// Add remote clusters declared in the Spec
+	for _, remoteCluster := range associatedEs.Spec.RemoteClusters {
+		if !remoteCluster.ElasticsearchRef.IsDefined() {
+			continue
+		}
+		esRef := remoteCluster.ElasticsearchRef.WithDefaultNamespace(associatedEs.Namespace)
+		expectedRemoteClusters[esRef.NamespacedName()] = struct{}{}
+	}
+
+	var list esv1.ElasticsearchList
+	if err := c.List(&list, &client.ListOptions{}); err != nil {
+		return nil, err
+	}
+
+	// Seek for Elasticsearch resources where this cluster is declared as a remote cluster
```

Comment:
I'm not sure I understand this bit. Why do we need to retrieve other references?

I thought it would be enough, when doing a reconciliation for `clusterA`, to reconcile all certificate secrets for `clusterA`. That is: certificates of remote clusters in the namespace of `clusterA` and certificates of `clusterA` in the remote clusters namespaces.

If `clusterB` declares `clusterA` as remote, secrets are reconciled in the `clusterB` association reconciliation. Why does reconciliation for `clusterA` need to care?
Aren't we reconciling the same secrets multiple times for different clusters?

Reply:
The reason is that `remoteClustersInvolvedWith` returns all the clusters for which an association has been established (This is needed to ensure that we remove all the secrets which are not needed anymore, not only the ones that are in the local namespace)

If we don't do the same here the remote CA will be deleted when the remote cluster is reconciled because it will not be expected: 

```golang
expectedRemoteClusters, err := getExpectedRemoteClusters(ctx, r.Client, localEs)
...
remoteClustersInvolved, err := remoteClustersInvolvedWith(ctx, r.Client, localClusterKey)
...
// Create or update expected remote CA
for remoteEsKey := range expectedRemoteClusters {
         ...
        delete(remoteClustersInvolved, remoteEsKey
}
// Delete existing but not expected remote CA
for toDelete := range remoteClustersInvolved { // Here
		results.WithError(deleteCertificateAuthorities(ctx, r, localClusterKey, toDelete))
}
```

If both clusters are reciprocally declared as "remote", it will induce an extra reconcile loop but since everything is already set and in cache I'm not sure that we should care too much.

Reply:
Got it. I guess the only way around would to label differently a CA that is a remote in another namespace and a CA that is a remote in the current namespace (it's very hard to explain this concept with words :()
Anyway I'm fine with the current approach, as you said an extra reconciliation on secrets we already have in the cache should not cost much.